### PR TITLE
Allows binding bean of inaccessible type

### LINKED ISSUE #242: @BindBean is incompatible with objects of non-public types
If `@BindBean` is specified with a type (let's call it `Foo`) that is a `public` `class` or `interface`, but is passed an object whose type (`ImplementsFoo`) is defined as a non-public `class`, the result is an exception in `BindBeanFactory`:

  `java.lang.IllegalAccessException: ... can not access a member of class`

This is because an attempt is made to invoke a method of a non-public class.  Note this exception is generated even though calling this method would generally be allowed as a function of polymorphism.

My take is that `BindBeanFactory` should use `setAccessible(true)` and allow the method call `readMethod.invoke(arg)` if the base type's method is public regardless whether the provided object's access modifier is public of not.

References:
1) [Relevant code](https://github.com/jdbi/jdbi/blob/e206c5681a68a5887beaecd71781dd8584b54332/src/main/java/org/skife/jdbi/v2/sqlobject/BindBeanFactory.java#L48) in `BindBeanFactory`.
2) [A more in depth discussion with examples](http://stackoverflow.com/questions/34009266/using-jdbi-bindbean-with-autovalue?noredirect=1#comment55774666_34009266) of how I discovered this issue while attempting to use `JDBI` with `AutoValue` on StackOverflow.

### PR DESCRIPTION
**Pull request by gsmith85**
Changes BinderFactory so that it allows the specification of the
Annotation type via a type token.

Changes the BindBean annotation such that a type can be provided as an
override.  This override will be used instead of the provided
argument's type for the purpose of binding.  This allows for
BindBeanFactory to access methods on args that are of types that are
inaccessible to the SqlObject package.

Fixes #242

### CODE REVIEW

File: src/main/java/org/skife/jdbi/v2/sqlobject/BindBean.java [hunk 1]
```diff
@@ -25,4 +25,13 @@
 {
     String BARE_BINDING = "___jdbi_bare___";
     String value() default BARE_BINDING;
+
+    // This override allows for explicitly specifying the type used to invoke the bound bean's methods.  By default the type is
+    // inferred from the argument that the annotated method is called with.  Providing an override may be required in certain
+    // situations, where the annotated parameter type is accessible from the org.skife.jdbi.v2.sqlobject package but a subtype
+    // is not.  The result of attempting to bind an inaccessible subtype being used without providing this override is a
+    // java.lang.IllegalAccessException.
+    Class<?> typeOverride() default Default.class;
```
Comment (gsmith85) [inline src/main/java/org/skife/jdbi/v2/sqlobject/BindBean.java]: Possibly `beanType` as a name instead of `typeOverride`?  I could go either way and am open to suggestions that would make the purpose of this parameter more explicit.

Comment (qualidafial) [inline src/main/java/org/skife/jdbi/v2/sqlobject/BindBean.java]: I'd rather just name this `type()` for brevity at the call site.

Also, can we tighten up this comment, and convert it into a javadoc? e.g.

    /**
     * The bean class of the argument. If omitted, <code>argument.getClass()</code> is used.
     */

File: src/main/java/org/skife/jdbi/v2/sqlobject/BinderFactory.java [hunk 2]
```diff
@@ -19,12 +19,12 @@
  * Factor for building {@link Binder} instances. This interface is used by
  * the {@link BindingAnnotation}
  */
-public interface BinderFactory
+public interface BinderFactory<T extends Annotation>
```
Comment (qualidafial) [inline src/main/java/org/skife/jdbi/v2/sqlobject/BinderFactory.java]: Can we move this generic parameter into a separate PR?

File: src/main/java/org/skife/jdbi/v2/sqlobject/BindBeanFactory.java [hunk 3]
```diff
@@ -40,7 +39,11 @@ public void bind(SQLStatement q, BindBean bind, Object arg)
                 }
 
                 try {
-                    BeanInfo infos = Introspector.getBeanInfo(arg.getClass());
+                    Class<?> beanType = bind.typeOverride().equals(BindBean.Default.class)
+                        ? arg.getClass()
+                        : bind.typeOverride();
+
+                    BeanInfo infos = Introspector.getBeanInfo(beanType);
```
Comment (qualidafial) [inline src/main/java/org/skife/jdbi/v2/sqlobject/BindBeanFactory.java]: Have you confirmed that the new tests fail without this code change?

**Comment by qualidafial** [comment]
@gsmith85 A few comments to take care of, then this should be good to merge.

File: src/main/java/org/skife/jdbi/v2/sqlobject/BindBeanFactory.java (hunk 3 above)
Comment (gsmith85) [inline src/main/java/org/skife/jdbi/v2/sqlobject/BindBeanFactory.java]: I have confirmed this, I wrote the test before the implementation.

**Comment by gsmith85** [comment]
I believe your changes have been addressed.  Let me know if you have any additional feedback!

**Comment by qualidafial** [comment]
Looks great. Thanks for the contribution!

**Comment by qualidafial** [comment]
@gsmith85 Heads up, you'll want to configure your name and contact email for future open source contributions:

    git config --global user.name "Greg Smith"
    git config --global user.email "your.email@example.com"

**Comment by gsmith85** [comment]
Yeah I just noticed I didn't update the author information on this commit too.  

Sorry about that.  Not much we can do now as updating the author would change the sha of the commit IIRC.

**Comment by stevenschlansker** [comment]
Cool.  This is a much better fix, thanks guys! :)
